crates/iota-grpc-api/src/lib.rs

```diff
@@ -0,0 +1,409 @@
+// Copyright (c) Mysten Labs, Inc.
+// Modifications Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{pin::Pin, sync::Arc};
+
+use tonic::{Request, Response, Status};
+pub mod checkpoint {
+    tonic::include_proto!("iota.grpc");
+}
+
+use checkpoint::checkpoint_service_server::CheckpointService;
+use iota_types::storage::RestStateReader;
+pub mod client;
+use bcs;
+use iota_types::{
+    full_checkpoint_content::CheckpointData,
+    grpc::{
+        CertifiedCheckpointSummary as GrpcCertifiedCheckpointSummary,
+        CheckpointData as GrpcCheckpointData,
+    },
+    messages_checkpoint::CertifiedCheckpointSummary,
+};
+use tracing::{debug, info};
+
+pub struct CheckpointGrpcService {
+    pub state_reader: Arc<dyn RestStateReader>,
+    pub grpc_checkpoint_summary_tx:
+        tokio::sync::broadcast::Sender<Arc<GrpcCertifiedCheckpointSummary>>,
+    pub grpc_checkpoint_data_tx: tokio::sync::broadcast::Sender<Arc<GrpcCheckpointData>>,
+}
+
+impl CheckpointGrpcService {
+    pub fn new(
+        state_reader: Arc<dyn RestStateReader>,
+        grpc_checkpoint_summary_tx: tokio::sync::broadcast::Sender<
+            Arc<GrpcCertifiedCheckpointSummary>,
+        >,
+        grpc_checkpoint_data_tx: tokio::sync::broadcast::Sender<Arc<GrpcCheckpointData>>,
+    ) -> Self {
+        Self {
+            state_reader,
+            grpc_checkpoint_summary_tx,
+            grpc_checkpoint_data_tx,
+        }
+    }
+}
+
+// Checkpoint stream item.
+// Note, checkpoint::Checkpoint may contain either checkpoint data or summary.
+type CheckpointStreamResult = Result<checkpoint::Checkpoint, Status>;
+
+// Helper trait for getting checkpoint data and summaries,
+// intended as an abstractoin for Arc<dyn RestStateReader>.
+trait CheckpointOracle<T>
+where
+    T: Send + Sync + 'static,
+    Self: Send + Sync + 'static,
+{
+    fn get_index(&self, item: &Arc<T>) -> u64;
+    fn ser(&self, item: &Arc<T>) -> Result<Vec<u8>, bcs::Error>;
```

Comment:
No need to save characters here, it is better to have readable code

Reply:
Renamed them in [9843a1e](https://github.com/iotaledger/iota/pull/7511/commits/9843a1e19a1e63af328c30932e510295ecdb413a).